File: src/main/java/apoc/util/QueueBasedSpliterator.java
package apoc.util;

import org.neo4j.procedure.TerminationGuard;

import java.util.Spliterator;
import java.util.concurrent.BlockingQueue;
import java.util.function.Consumer;

/**
 * @author mh
 * @since 06.12.17
 */
public class QueueBasedSpliterator<T> implements Spliterator<T> {
    private final BlockingQueue<T> queue;
    private T tombstone;
    private TerminationGuard terminationGuard;
    private boolean foundTombstone = false;

    public QueueBasedSpliterator(BlockingQueue<T> queue, T tombstone, TerminationGuard terminationGuard) {
        this.queue = queue;
        this.tombstone = tombstone;
        this.terminationGuard = terminationGuard;
    }

    @Override
    public boolean tryAdvance(Consumer<? super T> action) {
        if (Util.transactionIsTerminated(terminationGuard) || foundTombstone) return false;
        try {
            T element = queue.take();
            if (element.equals(tombstone)) {
                foundTombstone = true;
            } else {
                action.accept(element);
            }
            return true;
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }

    public Spliterator<T> trySplit() {
        return null;
    }

    public long estimateSize() {
        return Long.MAX_VALUE;
    }

    public int characteristics() {
        return NONNULL;
    }
}

File: extended/src/main/kotlin/apoc/kafka/consumer/procedures/QueueBasedSpliterator.kt
package apoc.kafka.consumer.procedures

import org.neo4j.graphdb.NotInTransactionException
import org.neo4j.graphdb.TransactionTerminatedException
import org.neo4j.procedure.TerminationGuard
import java.util.Spliterator
import java.util.concurrent.BlockingQueue
import java.util.concurrent.TimeUnit
import java.util.function.Consumer

/**
 * @author mh
 * @since 08.05.16 in APOC
 */
class QueueBasedSpliterator<T> constructor(private val queue: BlockingQueue<T>,
                                           private val tombstone: T,
                                           private val terminationGuard: TerminationGuard,
                                           private val timeout: Long = 10) : Spliterator<T?> {
    private var entry: T?

    init {
        entry = poll()
    }

    override fun tryAdvance(action: Consumer<in T?>): Boolean {
        if (transactionIsTerminated(terminationGuard)) return false
        if (isEnd) return false
        action.accept(entry)
        entry = poll()
        return !isEnd
    }

    private fun transactionIsTerminated(terminationGuard: TerminationGuard): Boolean {
        return try {
            terminationGuard.check()
            false
        } catch (e: Exception) {
            when (e) {
                is TransactionTerminatedException, is NotInTransactionException -> true
                else -> throw e
            }
        }
    }

    private val isEnd: Boolean
        private get() = entry == null || entry === tombstone

    private fun poll(): T? {
        return try {
            queue.poll(timeout, TimeUnit.SECONDS)
        } catch (e: InterruptedException) {
            null
        }
    }

    override fun trySplit(): Spliterator<T?>? {
        return null
    }

    override fun estimateSize(): Long {
        return Long.MAX_VALUE
    }

    override fun characteristics(): Int {
        return Spliterator.NONNULL
    }
}